File: src/main/java/net/onelitefeather/vulpes/backend/service/impl/NotificationServiceImpl.java
package net.onelitefeather.vulpes.backend.service.impl;

import io.micronaut.data.model.Page;
import io.micronaut.data.model.Pageable;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import net.onelitefeather.vulpes.api.model.NotificationEntity;
import net.onelitefeather.vulpes.api.repository.NotificationRepository;
import net.onelitefeather.vulpes.backend.domain.notification.NotificationModelDTO;
import net.onelitefeather.vulpes.backend.domain.notification.NotificationModelResponseDTO;
import net.onelitefeather.vulpes.backend.service.NotificationService;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Implementation of the NotificationService interface.
 */
@Singleton
public class NotificationServiceImpl implements NotificationService {

    private final NotificationRepository notificationRepository;

    @Inject
    public NotificationServiceImpl(NotificationRepository notificationRepository) {
        this.notificationRepository = notificationRepository;
    }

    @Override
    public NotificationModelResponseDTO.NotificationModelDTO createNotification(NotificationModelDTO notificationModelDTO) {
        NotificationEntity notificationModel = notificationModelDTO.toNotificationModel();
        NotificationEntity savedNotificationModel = notificationRepository.save(notificationModel);
        return NotificationModelResponseDTO.NotificationModelDTO.createDTO(savedNotificationModel);
    }

    @Override
    public NotificationModelResponseDTO updateNotification(NotificationModelDTO notificationModelDTO) {
        Optional<NotificationEntity> existingModel = notificationRepository.findById(notificationModelDTO.id());
        if (existingModel.isEmpty()) {
            return new NotificationModelResponseDTO.NotificationModelErrorDTO("Notification not found");
        }
        NotificationEntity notificationModel = notificationModelDTO.toNotificationModel();
        notificationModel = notificationRepository.update(notificationModel);
        return NotificationModelResponseDTO.NotificationModelDTO.createDTO(notificationModel);
    }

    @Override
    public NotificationModelResponseDTO deleteNotification(UUID id) {
        Optional<NotificationEntity> model = notificationRepository.findById(id);
        if (model.isPresent()) {
            notificationRepository.deleteById(id);
            return NotificationModelResponseDTO.NotificationModelDTO.createDTO(model.get());
        }
        return new NotificationModelResponseDTO.NotificationModelErrorDTO("Notification not found");
    }

    @Override
    public List<NotificationModelResponseDTO> deleteAllNotifications() {
        notificationRepository.deleteAll();
        return List.of();
    }

    @Override
    public Page<NotificationModelResponseDTO.NotificationModelDTO> getAllNotifications(Pageable pageable) {
        return notificationRepository.findAll(pageable).map(NotificationModelResponseDTO.NotificationModelDTO::createDTO);
    }

    @Override
    public Optional<NotificationEntity> findNotificationById(UUID id) {
        return notificationRepository.findById(id);
    }
}

File: src/main/kotlin/net/theevilreaper/vulpes/backend/service/dao/NotificationServiceImpl.kt
package net.theevilreaper.vulpes.backend.service.dao

import net.theevilreaper.vulpes.api.model.NotificationModel
import net.theevilreaper.vulpes.api.repository.NotificationRepository
import net.theevilreaper.vulpes.backend.dao.DatabaseAccessObject
import net.theevilreaper.vulpes.backend.exception.ResourceNotFoundException
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException

/**
 * @author theEvilReaper
 * @version 1.0.0
 * @since
 **/
@Service
class NotificationServiceImpl(val notificationRepository: NotificationRepository) :
    DatabaseAccessObject<NotificationModel> {

    override fun getAll(): ResponseEntity<List<NotificationModel>> {
        notificationRepository.findAll().let {
            return ResponseEntity.ok(it)
        }
    }

    override fun deleteAll(): ResponseEntity<List<NotificationModel>> {
        notificationRepository.deleteAll()
        return ResponseEntity.ok().build()
    }

    override fun delete(id: String): ResponseEntity<NotificationModel> {
        val notificationModel = notificationRepository.findById(id)
        if (notificationModel.isEmpty) {
            throw ResourceNotFoundException(HttpMethod.DELETE, id)
        }
        val model = notificationModel.get()
        notificationRepository.deleteById(model.id ?: id)
        return ResponseEntity.ok(model)
    }

    override fun add(model: NotificationModel): ResponseEntity<NotificationModel> {
        return ResponseEntity.ok(notificationRepository.save(model))
    }

    override fun update(model: NotificationModel): ResponseEntity<NotificationModel> {
        return ResponseEntity.ok(notificationRepository.save(model))
    }

    override fun getByID(id: String): ResponseEntity<NotificationModel> {
        val value = notificationRepository.findById(id)
        return if (value.isEmpty) throw ResourceNotFoundException(
            HttpMethod.GET, id
        ) else ResponseEntity.ok(value.get())
    }
}